### PR TITLE
Fix false positive dead catch on property assignment 

### PR DESCRIPTION
**Pull request by rajyan**
fixes https://github.com/phpstan/phpstan/issues/6256

I hope I understood the concept of ThrowPoint correctly...

### CODE REVIEW

**Comment by rajyan** [comment]
Rerunnig tests because the failing tests seemed not related to this PR.

**Comment by rajyan** [comment]
@ondrejmirtes 
Hi, this PR is ready for review.
This is my first PR, and I'm looking forward to contribute more!

File: src/Analyser/NodeScopeResolver.php [hunk 1]
```diff
@@ -3310,6 +3310,9 @@ private function processAssignVar(
 				if ($propertyReflection->canChangeTypeAfterAssignment()) {
 					$scope = $scope->assignExpression($var, $assignedExprType);
 				}
+				if ($assignedExprType->isSuperTypeOf($propertyReflection->getWritableType())->no()) {
+					$throwPoints[] = ThrowPoint::createImplicit($scope, $assignedExpr);
```
Comment (ondrejmirtes) [inline src/Analyser/NodeScopeResolver.php:3314]: The error might occur if we have a property with a native type `int` and we try to assign `int|string` into it. It means that the condition should be `!$assignedExprType->isSuperTypeOf($propertyReflection->getWritableType())->yes()` instead. Please add a test and modify the implementation.

Also there might be a situation where the property is not defined at all. In that case we should simulate MethodCall to `__set` and see what throw points that produces (when the method exists). You can do that by calling `$this->processExprNode(new MethodCall($var->var, '__set'), $scope, static function (): void {}, $context)->getThrowPoints()`.

There's a similar bug that you can look to in the next PR :) https://github.com/phpstan/phpstan/issues/4852

Thank you!

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

**Comment by rajyan** [comment]
@ondrejmirtes 
Thank you very much for your reviews.
I modified the tests and logic with your suggestions. (By the way, my `$assignedExprType->isSuperTypeOf($propertyReflection->getWritableType())` was checking in the opposite direction. The tests were useful to understand those cases.)

**Comment by rajyan** [comment]
@ondrejmirtes 
I'm looking deeper into throw points.
May I ask several questions?
1. Difference between explicit/implicit throw points
2. Do you think this case is a explicit one? (I think it can only throw `TypeError`)
3. If I should create an explicit throw point, is there a better way to get predefined classes type from scope than this?  
  *  `ThrowPoint::createExplicit($scope, $scope->getType(new New_(new Name(\TypeError::class))), $assignedExpr, false)`

**Comment by ondrejmirtes** [comment]
Instead of `$scope->getType(new New_(new Name(\TypeError::class)))` you can do `new ObjectType(\TypeError::class)`. See the documentation: https://phpstan.org/developing-extensions/type-system

Explicit/implicit throw points - "explicit" is for cases where the `@throws` tag is declared and we know the exact type that is thrown.
Implicit throw point is for cases where some kind of exception might be thrown but we're not sure which one. It's for cases where you're calling a function `foo()` and it has no `@throws` tag. Also see the blog post: https://phpstan.org/blog/bring-your-exceptions-under-control#what-does-absent-%40throws-above-a-function-mean%3F

So I guess in this case it's an explicit throw point.

**Comment by rajyan** [comment]
Thank you very much for your information!

> Instead of $scope->getType(new New_(new Name(\TypeError::class))) you can do new ObjectType(\TypeError::class).

https://github.com/phpstan/phpstan-src/blob/d32942a3158a26f08252638dc660467292bbf881/src/Type/ObjectType.php#L61
Thanks, I had a misunderstanding watching property `$superTypes` here. I was thinking that  calling `new ObjectType(\TypeError::class)` is not setting the `$superTypes`, but this property was only a cache.

File: src/Analyser/NodeScopeResolver.php [hunk 2]
```diff
@@ -3314,11 +3316,22 @@ private function processAssignVar(
...
+					$throwPoints[] = ThrowPoint::createExplicit($scope, new ObjectType(TypeError::class), $assignedExpr, false);
+				}
 			} else {
 				// fallback
 				$assignedExprType = $scope->getType($assignedExpr);
 				$nodeCallback(new PropertyAssignNode($var, $assignedExpr, $isAssignOp), $scope);
 				$scope = $scope->assignExpression($var, $assignedExprType);
+				// simulate dynamic property assign to get throw points
+				$throwPoints = array_merge($throwPoints, $this->processExprNode(
+					new MethodCall($var->var, '__set'),
```
Comment (ondrejmirtes) [inline src/Analyser/NodeScopeResolver.php:3329]: This needs some tests :)

1) Class without the `__set` magic method - does not throw anything (dead catch reported).
2) Class with `__set` magic method and `@throws void` annotation (dead catch reported).
3) Class with `__set` magic method and `@throws SomeException` (dead catch not reported).

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Analyser/NodeScopeResolver.php (hunk 2 above)
Comment (rajyan) [reply src/Analyser/NodeScopeResolver.php:3329]: :+1: 

For the first case, there is a excplitThrowPoint already, when method reflection is not found.
I'm not sure with the purpose of this throw point, so I'm gonna check it.
https://github.com/phpstan/phpstan-src/blob/21674384d26e6680d98545db11a6dfa0e12eb4fb/src/Analyser/NodeScopeResolver.php#L1994

Comment (rajyan) [reply src/Analyser/NodeScopeResolver.php:3329]: It was from this case
https://github.com/phpstan/phpstan-src/commit/84b09349dc04fa30097900b5deb8e316b1657e4b

Comment (rajyan) [reply src/Analyser/NodeScopeResolver.php:3328]: Ok, understood
I shouldn’t process `__set` if its not declared.

**Review by ondrejmirtes — APPROVED** [review]

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**